tidy3d/components/eme/grid.py

```diff
@@ -116,6 +141,46 @@ def make_grid(self, center: Coordinate, size: Size, axis: Axis) -> EMEGrid:
             by the EME grid spec.
         """
 
+    @property
+    def real_cell_indices(self) -> int:
+        """The cell indices inside this EME grid, starting at 0,
+        not including periodic repetition of cells."""
+        return np.arange(self.num_real_cells)
+
+    @property
+    @abstractmethod
+    def num_real_cells(self) -> int:
+        """Number of real cells in this EME grid spec."""
+
+    @property
+    def virtual_cell_indices(self) -> int:
+        """The cell indices inside this EME grid, starting at 0
+        and including periodic repetition of cells with ``num_reps``."""
+        return list(self.real_cell_indices) * self.num_reps
+
+    @property
+    def num_virtual_cells(self) -> int:
+        """Number of virtual cells in this EME grid spec."""
+        return len(self.virtual_cell_indices)
+
+    def updated_copy_num_reps(self, num_reps: Dict[str, pd.PositiveInt]) -> EMEGridSpec:
+        """Update ``num_reps`` of named subgrids."""
+        if self.name is not None:
+            new_num_reps = num_reps.get(self.name)
+            if new_num_reps is not None:
+                return self.updated_copy(num_reps=new_num_reps)
+        return self
+
+    @property
+    def cell_index_pairs(self) -> List[pd.NonNegativeInt]:
```

Comment:
Should this and `updated_copy_num_reps` be private? Or is there a reason we might want them user-exposed?

Reply:
Yeah, fair